pkg/storage/ingest/writer.go

```diff
@@ -0,0 +1,299 @@
+		// because we expect the producer to be high volume and in such case the library doc recommend to not
+		// set linger.
+		kgo.ProducerBatchMaxBytes(16_000_000),
+
+		// A cluster metadata update is a request sent to a broker and getting back the map of partitions and
+		// the leader broker for each partition. The cluster metadata can be updated (a) periodically or
+		// (b) when some events occur (e.g. backoff due to errors).
+		//
+		// MetadataMinAge() sets the minimum time between two cluster metadata updates due to events.
+		// MetadataMaxAge() sets how frequently the periodic update should occur.
+		//
+		// We configure the client to frequently refresh the cluster metadata. This allows us to quickly react in case
+		// some brokers are unhealthy.
+		//
+		// The side effect of frequently refreshing the cluster metadata is that if the backend returns each time a
+		// different authoritative owner for a partition, then each time cluster metadata is updated the Kafka client
+		// will client a new connection for each partition, leading to an high connections churn rate.
+		kgo.MetadataMaxAge(10*time.Second),
+
+		// By default, the Kafka client allows 1 Produce in-flight request per broker. Disabling write idempotency
+		// (which we don't need), we can increase the max number of in-flight Produce requests per broker. An higher
+		// number of in-flight requests, in addition to short buffering ("linger") in client side before firing the
+		// next Produce request allows us to reduce the end-to-end latency.
+		//
+		// The result of the multiplication of producer linger and max inflight requests should match the maximum
+		// Produce latency expected by the Kafka backend in a steady state. For example, 50ms * 20 requests = 1s,
+		// which means the Kafka client will keep issuing a Produce request every 50ms as far as the Kafka backend
+		// doesn't take longer than 1s to process them (if it takes longer, the client will buffer data and stop
+		// issuing new Produce requests until some previous ones complete).
+		kgo.DisableIdempotentWrite(),
+		kgo.ProducerLinger(50*time.Millisecond),
+		kgo.MaxProduceRequestsInflightPerBroker(w.writerCfg.kafkaMaxInflightProduceRequests),
+
+		// Unlimited number of Produce retries but a deadline on the max time a record can take to be delivered.
+		// With the default config it would retry infinitely.
+		//
+		// Details of the involved timeouts:
+		// - RecordDeliveryTimeout: how long a Kafka client Produce() call can take for a given record. The overhead
+		//   timeout is NOT applied.
+		// - ProduceRequestTimeout: how a long to wait the response for the Produce request (the Kafka protocol message)
+		//   after being sent on the network. The actual timeout is increased by the configured overhead.
+		kgo.RecordRetries(math.MaxInt64),
+		kgo.RecordDeliveryTimeout(w.writerCfg.KafkaWriteTimeout),
+		kgo.ProduceRequestTimeout(w.writerCfg.KafkaWriteTimeout),
```

Comment:
_Discussed offline._

The TL;DR is that you depends on the actual failure scenario.

If the backend is slow, then an higher `ProduceRequestTimeout` increases the chances of a successful request (in other words, one try of 10s is better than 2x5s tries because maybe just waiting longer fixes it).

If the backend is unhealthy, then a shorter `ProduceRequestTimeout` is better because it will retry within the `RecordDeliveryTimeout` but the retry will be successful only if the meanwhile the cluster metadata has been updated and the replica owning a given partition has been moved to another broker, otherwise the client will just keep trying to connect to the previous (unhealthy) one. In a setup where these timeouts are relatively low, it may not be that common that the unhealthy replica has been actually detected as unhealthy and the replica owner for a given partition moved.